From 1a8dc5200c77291e69279ea458930bf696fc6994 Mon Sep 17 00:00:00 2001
From: Howard Hsu <howard-yh.hsu@mediatek.com>
Date: Thu, 10 Jul 2025 19:26:05 +0800
Subject: [PATCH 105/123] mtk: mt76: mt7996: add support configure critical
 packet mode by debugfs

Add new debugfs cp_mode to configure critical packet mode. This debugfs
can decide the ac value for priority packet. This debugfs is needed for
QoS management program.

Usage:
echo val > /sys/kernel/debug/ieee80211/phy0/mt76/cp_mode

val:
0x0: BK
0x1: BE
0x2: VI
0x3: VO

Signed-off-by: Howard Hsu <howard-yh.hsu@mediatek.com>
---
 mt7996/mtk_debugfs_i.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/mt7996/mtk_debugfs_i.c b/mt7996/mtk_debugfs_i.c
index 98002279..d58ea794 100644
--- a/mt7996/mtk_debugfs_i.c
+++ b/mt7996/mtk_debugfs_i.c
@@ -1256,6 +1256,23 @@ mt7996_mru_probe_enable_set(void *data, u64 val)
 DEFINE_DEBUGFS_ATTRIBUTE(fops_mru_probe_enable, mt7996_mru_probe_enable_get,
 			 mt7996_mru_probe_enable_set, "%lld\n");
 
+static int
+mt7996_cp_mode_set(void *data, u64 val)
+{
+	struct mt7996_dev *dev = data;
+
+	/* mt7996 does not support configuring the critical packet mode
+	 * due to hardware limitations.
+	 */
+	if (is_mt7996(&dev->mt76) || val > IEEE80211_AC_BK)
+		return -EINVAL;
+
+	return mt7996_mcu_cp_support(dev, val);
+}
+
+DEFINE_DEBUGFS_ATTRIBUTE(fops_cp_mode, NULL,
+			 mt7996_cp_mode_set, "%llu\n");
+
 int mt7996_mtk_init_dev_debugfs_internal(struct mt7996_phy *phy, struct dentry *dir)
 {
 	struct mt7996_dev *dev = phy->dev;
@@ -1276,6 +1293,7 @@ int mt7996_mtk_init_dev_debugfs_internal(struct mt7996_phy *phy, struct dentry *
 
 	debugfs_create_u8("dump_ple_txd", 0600, dir, &dev->dbg.dump_ple_txd);
 	debugfs_create_file("txop", 0600, dir, dev, &fops_mt7996_txop);
+	debugfs_create_file("cp_mode", 0200, dir, dev, &fops_cp_mode);
 
 	/* MLO related Table */
 	debugfs_create_file("mat_table", 0400, dir, dev, &mt7996_mat_table_fops);
-- 
2.45.2

